File: sync-server/rcdrmd-sync-server.js
/**
 * RcdRmd - Local Cross-Browser Sync Server (Node.js)
 * Coordinates Bilibili tabs, workspaces & deleted video records across Edge, Chrome, and Canary.
 * Zero external dependencies: uses built-in 'http' and 'fs'.
 */

const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 32188;
const DATASET_FILE = path.join(__dirname, 'rcdrmd-dataset.json');

// Shared token that must accompany every data request.
// The real hardening is localhost-only binding; this token blocks casual
// drive-by requests from arbitrary web pages that don't know it.
const SYNC_TOKEN = 'rcdrmd-local-sync-token';

function isValidToken(req) {
  return req.headers['x-rcdrmd-token'] === SYNC_TOKEN;
}

function loadDataset() {
  try {
    if (fs.existsSync(DATASET_FILE)) {
      const content = fs.readFileSync(DATASET_FILE, 'utf-8');
      return JSON.parse(content);
    }
  } catch (err) {
    console.error('[Sync Server] Error reading dataset:', err.message);
  }

  return {
    version: '1.0.0',
    updatedAt: new Date().toISOString(),
    browsers: {},
    deletedVideos: [],
    allTrackedHistory: {}
  };
}

function saveDataset(data) {
  try {
    data.updatedAt = new Date().toISOString();
    const tempFile = `${DATASET_FILE}.tmp`;
    fs.writeFileSync(tempFile, JSON.stringify(data, null, 2), 'utf-8');
    fs.renameSync(tempFile, DATASET_FILE);
  } catch (err) {
    console.error('[Sync Server] Error saving dataset:', err.message);
  }
}

let dataset = loadDataset();

function setCorsHeaders(req, res) {
  // Only reflect the origin when the caller presents a valid token.
  // The extension talks to this server via its `host_permissions`, so it
  // bypasses CORS entirely and is unaffected by this tightening.
  if (isValidToken(req)) {
    res.setHeader('Access-Control-Allow-Origin', '*');
  }
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Accept, X-RcdRmd-Token');
}

const server = http.createServer((req, res) => {
  setCorsHeaders(req, res);

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  const url = new URL(req.url, `http://localhost:${PORT}`);

  // Authentication check: require the shared token on every data endpoint.
  // Without it, an arbitrary web page cannot read the dataset or inject
  // polluting records via a no-cors form/body POST.
  if ((req.method === 'GET' || req.method === 'POST') && !isValidToken(req)) {
    res.writeHead(401, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Unauthorized: missing or invalid X-RcdRmd-Token' }));
    return;
  }

  // Route: GET /api/status
  if (req.method === 'GET' && url.pathname === '/api/status') {
    const activeBrowsers = Object.keys(dataset.browsers || {});
    let totalTabs = 0;
    activeBrowsers.forEach(b => {
      totalTabs += (dataset.browsers[b].tabs || []).length;
    });

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
      status: 'online',
      version: '1.0.0',
      port: PORT,
      activeBrowsers,
      totalTabsAcrossBrowsers: totalTabs,
      totalDeletedVideos: (dataset.deletedVideos || []).length,
      datasetPath: DATASET_FILE,
      updatedAt: dataset.updatedAt
    }));
    return;
  }

  // Route: GET /api/dataset
  if (req.method === 'GET' && url.pathname === '/api/dataset') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(dataset, null, 2));
    return;
  }

  // Route: POST /api/sync
  if (req.method === 'POST' && url.pathname === '/api/sync') {
    let body = '';
    req.on('data', chunk => {
      body += chunk;
      if (body.length > 10 * 1024 * 1024) {
        req.destroy();
      }
    });

    req.on('end', () => {
      try {
        const payload = JSON.parse(body || '{}');
        const browserName = payload.browser || 'Unknown Browser';
        const incomingTabs = Array.isArray(payload.tabs) ? payload.tabs : [];
        const incomingDeleted = Array.isArray(payload.deletedVideos) ? payload.deletedVideos : [];
        const incomingWindows = payload.windows || {};

        // 1. Update this browser's tabs & workspaces
        if (!dataset.browsers) dataset.browsers = {};
        dataset.browsers[browserName] = {
          name: browserName,
          lastHeartbeat: new Date().toISOString(),
          windows: incomingWindows,
          tabs: incomingTabs
        };

        // 2. Cache tracked video titles in history
        if (!dataset.allTrackedHistory) dataset.allTrackedHistory = {};
        incomingTabs.forEach(t => {
          if (t.bvid && t.title) {
            dataset.allTrackedHistory[t.bvid] = {
              title: t.title,
              channel: t.channel || '',
              browser: browserName,
              workspace: t.workspace || '',
              lastSeen: new Date().toISOString()
            };
          }
        });

        // 3. Merge deleted videos
        if (!dataset.deletedVideos) dataset.deletedVideos = [];
        incomingDeleted.forEach(delItem => {
          const matchIdx = dataset.deletedVideos.findIndex(v =>
            (v.bvid && delItem.bvid && v.bvid === delItem.bvid) ||
            (v.id && delItem.id && v.id === delItem.id)
          );

          if (matchIdx === -1) {
            dataset.deletedVideos.unshift({
              ...delItem,
              recordedByServerAt: new Date().toISOString()
            });
            console.log(`[Sync Server] ⚠️ Registered DELETED video from ${browserName}: "${delItem.title}" (${delItem.channel})`);
          } else {
            if (delItem.title && delItem.title !== '已删除的B站视频' && delItem.title !== 'Deleted Bilibili Video') {
              dataset.deletedVideos[matchIdx].title = delItem.title;
            }
            if (delItem.channel) {
              dataset.deletedVideos[matchIdx].channel = delItem.channel;
            }
          }
        });

        saveDataset(dataset);

        // 4. Flatten all opening tabs from all browsers
        const allOpeningTabs = [];
        Object.keys(dataset.browsers).forEach(bKey => {
          const bData = dataset.browsers[bKey];
          if (Array.isArray(bData.tabs)) {
            allOpeningTabs.push(...bData.tabs);
          }
        });

        console.log(`[Sync Server] Synced with "${browserName}" (${incomingTabs.length} tabs, ${incomingDeleted.length} deleted). Total across browsers: ${allOpeningTabs.length} tabs.`);

        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({
          success: true,
          serverTime: new Date().toISOString(),
          activeBrowsers: Object.keys(dataset.browsers),
          allOpeningTabs: allOpeningTabs,
          deletedVideos: dataset.deletedVideos
        }));
      } catch (err) {
        console.error('[Sync Server] Error processing sync request:', err);
        res.writeHead(400, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: err.message }));
      }
    });
    return;
  }

  res.writeHead(404, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ error: 'Endpoint not found' }));
});

server.listen(PORT, '127.0.0.1', () => {
  console.log('====================================================');
  console.log(`  RcdRmd Local Cross-Browser Sync Hub`);
  console.log(`  Running on: http://127.0.0.1:${PORT}`);
  console.log(`  Dataset storage: ${DATASET_FILE}`);
  console.log(`  Ready for Microsoft Edge, Chrome & Chrome Canary`);
  console.log('====================================================');
});

File: sync-server/rcdrmd-sync-server.py
"""
RcdRmd - Local Cross-Browser Sync Server (Python)
Coordinates Bilibili tabs, workspaces & deleted video records across Edge, Chrome, and Canary.
Zero external dependencies: uses Python standard library http.server.
"""

import json
import os
import sys
from datetime import datetime
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse

PORT = 32188
DATASET_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'rcdrmd-dataset.json')

def load_dataset():
    if os.path.exists(DATASET_FILE):
        try:
            with open(DATASET_FILE, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as e:
            print(f"[Sync Server] Error loading dataset: {e}")
    return {
        "version": "1.0.0",
        "updatedAt": datetime.utcnow().isoformat() + "Z",
        "browsers": {},
        "deletedVideos": [],
        "allTrackedHistory": {}
    }

def save_dataset(data):
    try:
        data["updatedAt"] = datetime.utcnow().isoformat() + "Z"
        temp_file = DATASET_FILE + ".tmp"
        with open(temp_file, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        os.replace(temp_file, DATASET_FILE)
    except Exception as e:
        print(f"[Sync Server] Error saving dataset: {e}")

dataset = load_dataset()

class SyncHandler(BaseHTTPRequestHandler):
    def _set_cors(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Accept')

    def do_OPTIONS(self):
        self.send_response(204)
        self._set_cors()
        self.end_headers()

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == '/api/status':
            self.send_response(200)
            self._set_cors()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            active_browsers = list(dataset.get("browsers", {}).keys())
            total_tabs = sum(len(b.get("tabs", [])) for b in dataset.get("browsers", {}).values())
            res = {
                "status": "online",
                "version": "1.0.0",
                "port": PORT,
                "activeBrowsers": active_browsers,
                "totalTabsAcrossBrowsers": total_tabs,
                "totalDeletedVideos": len(dataset.get("deletedVideos", [])),
                "datasetPath": DATASET_FILE,
                "updatedAt": dataset.get("updatedAt")
            }
            self.wfile.write(json.dumps(res, ensure_ascii=False).encode('utf-8'))
            return

        if parsed.path == '/api/dataset':
            self.send_response(200)
            self._set_cors()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps(dataset, ensure_ascii=False, indent=2).encode('utf-8'))
            return

        self.send_response(404)
        self._set_cors()
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(b'{"error":"Not Found"}')

    def do_POST(self):
        parsed = urlparse(self.path)
        if parsed.path == '/api/sync':
            length = int(self.headers.get('Content-Length', 0))
            body_bytes = self.rfile.read(length)
            try:
                payload = json.loads(body_bytes.decode('utf-8'))
                browser_name = payload.get("browser", "Unknown Browser")
                incoming_tabs = payload.get("tabs", [])
                incoming_deleted = payload.get("deletedVideos", [])
                incoming_windows = payload.get("windows", {})

                if "browsers" not in dataset:
                    dataset["browsers"] = {}
                dataset["browsers"][browser_name] = {
                    "name": browser_name,
                    "lastHeartbeat": datetime.utcnow().isoformat() + "Z",
                    "windows": incoming_windows,
                    "tabs": incoming_tabs
                }

                if "allTrackedHistory" not in dataset:
                    dataset["allTrackedHistory"] = {}
                for t in incoming_tabs:
                    bvid = t.get("bvid")
                    title = t.get("title")
                    if bvid and title:
                        dataset["allTrackedHistory"][bvid] = {
                            "title": title,
                            "channel": t.get("channel", ""),
                            "browser": browser_name,
                            "workspace": t.get("workspace", ""),
                            "lastSeen": datetime.utcnow().isoformat() + "Z"
                        }

                if "deletedVideos" not in dataset:
                    dataset["deletedVideos"] = []
                for del_item in incoming_deleted:
                    r = (del_item.get("reason") or "").lower()
                    if "http 412" in r or "http 403" in r or "http 429" in r or "http 5" in r:
                        continue
                    bvid = del_item.get("bvid")
                    item_id = del_item.get("id")
                    match_idx = -1
                    for idx, v in enumerate(dataset["deletedVideos"]):
                        if (bvid and v.get("bvid") == bvid) or (item_id and v.get("id") == item_id):
                            match_idx = idx
                            break

                    if match_idx == -1:
                        del_item_copy = dict(del_item)
                        del_item_copy["recordedByServerAt"] = datetime.utcnow().isoformat() + "Z"
                        dataset["deletedVideos"].insert(0, del_item_copy)
                        print(f"[Sync Server] ⚠️ Registered DELETED video from {browser_name}: {del_item.get('title')}")
                    else:
                        if del_item.get("title") and del_item.get("title") not in ["已删除的B站视频", "Deleted Bilibili Video"]:
                            dataset["deletedVideos"][match_idx]["title"] = del_item["title"]
                        if del_item.get("channel"):
                            dataset["deletedVideos"][match_idx]["channel"] = del_item["channel"]

                dataset["deletedVideos"] = [
                    v for v in dataset["deletedVideos"]
                    if "http 412" not in (v.get("reason") or "").lower()
                    and "http 403" not in (v.get("reason") or "").lower()
                    and "http 429" not in (v.get("reason") or "").lower()
                ]

                save_dataset(dataset)

                all_tabs = []
                for b_data in dataset.get("browsers", {}).values():
                    all_tabs.extend(b_data.get("tabs", []))

                print(f"[Sync Server] Synced with '{browser_name}' ({len(incoming_tabs)} tabs). Total: {len(all_tabs)} tabs.")

                self.send_response(200)
                self._set_cors()
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                res = {
                    "success": True,
                    "serverTime": datetime.utcnow().isoformat() + "Z",
                    "activeBrowsers": list(dataset.get("browsers", {}).keys()),
                    "allOpeningTabs": all_tabs,
                    "deletedVideos": dataset.get("deletedVideos", [])
                }
                self.wfile.write(json.dumps(res, ensure_ascii=False).encode('utf-8'))
                return
            except Exception as e:
                print(f"[Sync Server] Sync error: {e}")
                self.send_response(400)
                self._set_cors()
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({"error": str(e)}).encode('utf-8'))
                return

        self.send_response(404)
        self._set_cors()
        self.end_headers()

def run():
    server_address = ('127.0.0.1', PORT)
    httpd = HTTPServer(server_address, SyncHandler)
    print("=" * 52)
    print(f"  RcdRmd Local Cross-Browser Sync Hub (Python)")
    print(f"  Running on: http://127.0.0.1:{PORT}")
    print(f"  Dataset storage: {DATASET_FILE}")
    print(f"  Ready for Microsoft Edge, Chrome & Chrome Canary")
    print("=" * 52)
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n[Sync Server] Shutting down.")
        httpd.server_close()

if __name__ == '__main__':
    run()
